src/eth_provider/database/ethereum.rs

```diff
@@ -356,4 +359,190 @@ mod tests {
+        assert_eq!(database.header(B256::from(U256::from(rng.gen::<u64>())).into()).await.unwrap(), None);
+
+        // Test retrieving non-existing header by block number
+        assert_eq!(database.header(rng.gen::<u64>().into()).await.unwrap(), None);
+    }
+
+    async fn test_get_blocks(database: &Database, mongo_fuzzer: &MongoFuzzer, u: &mut arbitrary::Unstructured<'_>) {
+        let header = mongo_fuzzer
+            .documents()
+            .get(&CollectionDB::Headers)
+            .unwrap()
+            .first()
+            .unwrap()
+            .extract_stored_header()
+            .unwrap()
+            .header
+            .clone();
+
+        let block_hash = header.hash.unwrap();
+
+        assert!(database
+            .block(block_hash.into(), true)
+            .await
+            .unwrap()
+            .unwrap()
+            .inner
+            .transactions
+            .as_transactions()
+            .is_some());
+
+        assert!(database
+            .block(block_hash.into(), false)
+            .await
+            .unwrap()
+            .unwrap()
+            .inner
+            .transactions
+            .as_hashes()
+            .is_some());
+
+        let first_block_number = header.number.unwrap();
+
+        assert!(database
+            .block(first_block_number.into(), true)
+            .await
+            .unwrap()
+            .unwrap()
+            .inner
+            .transactions
+            .as_transactions()
+            .is_some());
+
+        assert!(database
+            .block(first_block_number.into(), false)
+            .await
+            .unwrap()
+            .unwrap()
+            .inner
+            .transactions
+            .as_hashes()
+            .is_some());
+
+        let mut rng = rand::thread_rng();
+
+        // Test retrieving non-existing block by block hash
+        assert_eq!(database.block(B256::from(U256::from(rng.gen::<u64>())).into(), false).await.unwrap(), None);
+
+        // Test retrieving non-existing block by block number
+        assert_eq!(database.block(rng.gen::<u64>().into(), false).await.unwrap(), None);
+
+        // test withdrawals_root raises an error
+        let block_number: u64 = 2;
+
+        let faulty_header = Header {
+            hash: Some(B256::arbitrary(u).unwrap()),
+            total_difficulty: Some(U256::arbitrary(u).unwrap()),
+            mix_hash: Some(B256::arbitrary(u).unwrap()),
+            nonce: Some(B64::arbitrary(u).unwrap()),
+            withdrawals_root: Some(rng.gen::<B256>()),
+            base_fee_per_gas: Some(u128::from(u64::arbitrary(u).unwrap())),
+            blob_gas_used: Some(u128::from(u64::arbitrary(u).unwrap())),
+            excess_blob_gas: Some(u128::from(u64::arbitrary(u).unwrap())),
+            gas_limit: u128::from(u64::arbitrary(u).unwrap()),
+            gas_used: u128::from(u64::arbitrary(u).unwrap()),
+            number: Some(block_number),
+            ..Header::arbitrary(u).unwrap()
+        };
```

Comment:
why can't you use `StoredHeader::arbitrary_with_optional_fields` and then modify the block number field? I think it's cleaner than reimplementing what is already here 

https://github.com/kkrt-labs/kakarot-rpc/blob/0e759fda9fc77e3b22f55cfbefc3b35cef6d4855/src/eth_provider/database/types/header.rs#L17-L37

Reply:
Because it hard codes the withdrawal root with an empty hash and I'm testing it when it the withdrawal root is not an empty hash and just it errors out

Reply:
you could do something around:

```rust
let mut my_header = StoredHeader:: arbitrary_with_optional_fields
my_header.header. withdrawals_root = 
my_header.header. number = 
```